firmware/lib/Gantry/Axis.cpp

```diff
@@ -1,311 +1,569 @@
-void TC3_Handler(void)
-{   
-    // DEBUG_PRINT_VAL("Desired count", axis_x.encoder.desired);
-    // DEBUG_PRINT_VAL("Current count", axis_x.encoder.current);
-    TC_GetStatus(TC1, 0); // Get current status on the selected channel
-    PIOC->PIO_ODSR ^= PIO_ODSR_P25;
-    reset_timer(TC1, 0, TC3_IRQn, axis_x.vel); 
+/**
+ * @brief Common FAR limit switch ISR handler
+ * 
+ * @param axis Pointer to the Axis whose limit switch triggered the interrupt
+ */
+static __attribute__((always_inline)) inline void handle_isr_ls_far(Axis *axis)
+{
+    stop_axis(axis);
+    // Limit switch reads LOW when pressed
+    axis->state.ls_far_pressed = (digitalRead(axis->io.pin_ls_far) == LOW);
 }
 
-// isr for operating x axis motor acceleration 
-void TC4_Handler(void)
-{   
-    // DEBUG_PRINT_VAL("Desired count", axis_x.encoder.desired);
-    // DEBUG_PRINT_VAL("Current count", axis_x.encoder.current);
-    TC_GetStatus(TC1, 1); // Get current status on the selected channel
-    // first accelerate
-    if (axis_x.tragectory_segment == VEL_SEG_ACCELERATE) {
-        if (((axis_x.encoder.current < axis_x.encoder.desired) != axis_x.dir) && axis_x.vel < axis_x.vel_max) {
-            axis_x.vel++;
-        }
-        else {
-            axis_x.tragectory_segment = VEL_SEG_HOLD;
-            int delta = axis_x.encoder.desired - axis_x.encoder.current;
-            axis_x.encoder.desired = axis_x.encoder.current + axis_x.vel_profile_cur_trap[1] + delta;
-            // DEBUG_PRINT_VAL("accel d-c", delta);
-        }
+static __attribute__((always_inline)) inline void handle_isr_step(Axis *axis)
+{
+    // Reset the timer with a new velocity if it has changed
+    if (axis->state.next_velocity != axis->state.velocity) {
+        axis->state.velocity = axis->state.next_velocity;
+        reset_pwm_timer(
+            axis->io.tc_step,
+            axis->io.tc_step_channel,
+            axis->io.tc_step_irq,
+            axis->state.velocity,
+            VEL_DUTY_CYCLE);
     }
-    // hold velocity
-    else if (axis_x.tragectory_segment == VEL_SEG_HOLD) {
-        if ((axis_x.encoder.current < axis_x.encoder.desired) != axis_x.dir) {
-            // DEBUG_PRINT_VAL("DC H", axis_x.encoder.desired);
-            // DEBUG_PRINT_VAL("CC H", axis_x.encoder.current);
-        }
-        else {
-            axis_x.tragectory_segment = VEL_SEG_DECELERATE;
-            int delta = axis_x.encoder.desired - axis_x.encoder.current;
-            axis_x.encoder.desired = axis_x.encoder.current + axis_x.vel_profile_cur_trap[2] + delta;
-            // DEBUG_PRINT_VAL("hold d-c", delta);
+}
+
+/**
+ * @brief Common acceleration timer ISR handler
+ * 
+ * @param axis Pointer to the Axis whose acceleration timer triggered the interrupt
+ */
+static __attribute__((always_inline)) inline void handle_isr_accel(Axis *axis)
+{
+    if (!axis->state.moving) return;
+
+    switch (axis->state.velocity_segment) {
+        case VEL_SEG_ACCELERATE:
```

Comment:
switch statements will definitely be faster for 5+ items because they will get compiled into a jump table. For less than 5 items there is negligible difference, so I went with switch cause it seemed more legible